lib/internal/fs/recursive_watch.js

```diff
@@ -0,0 +1,290 @@
+    if (this.#watchingFile) {
+      return StringPrototypeSlice(this.#rootPath, StringPrototypeLastIndexOf(this.#rootPath, '/') + 1);
+    }
+
+    // When watching directories, and root directory is changed, return the full file path
+    if (file === this.#rootPath) {
+      return this.#rootPath;
+    }
+
+    // Otherwise return relative path for all files & folders
+    return path.relative(this.#rootPath, file);
+  }
+
+  #unwatchFiles(file) {
+    const { unwatchFile } = lazyLoadFsSync();
+
+    for (const filename of this.#files.keys()) {
+      if (StringPrototypeStartsWith(filename, file)) {
+        unwatchFile(filename);
+      }
+    }
+  }
+
+  async #watchFolder(folder) {
+    const { opendir } = lazyLoadFsPromises();
+
+    try {
+      const files = await opendir(folder);
+
+      for await (const file of files) {
+        if (this.#closed) {
+          break;
+        }
+
+        const f = path.join(folder, file.name);
+
+        if (!this.#files.has(f)) {
+          this.emit('change', 'rename', this.#getPath(f));
+
+          if (file.isFile()) {
+            this.#watchFile(f);
+          } else {
+            this.#files.set(f, file);
+
+            if (file.isDirectory() && !file.isSymbolicLink()) {
+              await this.#watchFolder(f);
+            }
+          }
+        }
+      }
+    } catch (error) {
+      this.emit('error', error);
+    }
+  }
+
+  #watchFile(file) {
+    if (this.#closed) {
+      return;
+    }
+
+    const { watchFile } = lazyLoadFsSync();
+    const existingStat = this.#files.get(file);
+
+    watchFile(file, {
+      persistent: this.#options.persistent,
+    }, (statWatcher, previousStatWatcher) => {
+      if (existingStat && !existingStat.isDirectory() &&
+        statWatcher.nlink !== 0 && existingStat.mtime?.getTime() === statWatcher.mtime?.getTime()) {
+        return;
+      }
+
+      this.#files.set(file, statWatcher);
+
+      if (statWatcher.birthtimeMs === 0 && previousStatWatcher.birthtimeMs !== 0) {
+        // The file is now deleted
+        this.#files.delete(file);
+        this.emit('change', 'rename', this.#getPath(file));
+        this.#unwatchFiles(file);
+      } else {
+        this.emit('change', 'change', this.#getPath(file));
+
+        if (statWatcher.isDirectory() && !statWatcher.isSymbolicLink()) {
+          this.#watchFolder(file);
+        }
+      }
+    });
+  }
+
+  [kFSWatchStart](filename) {
+    filename = path.resolve(getValidatedPath(filename));
+
+    try {
+      const file = lazyLoadFsSync().statSync(filename);
+
+      this.#rootPath = filename;
+      this.#closed = false;
+      this.#watchingFile = file.isFile();
+
+      if (file.isDirectory()) {
+        this.#files.set(filename, file);
+
+        PromisePrototypeThen(
+          traverse(filename, this.#files),
+          () => {
+            for (const f of this.#files.keys()) {
+              this.#watchFile(f);
+            }
+          },
+        );
```

Comment:
Those promises are never awaited and have no "catch handler", meaning that if any promise rejects for any reason, the program would crash without recourse for the user. Can we make sure we can await them at least in the async version?